clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +68,724 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+        let body = cx.tcx.optimized_mir(def_id.to_def_id());
+
+        let mut has_clone = false;
+        let mut clone_calls = BitSet::new_empty(body.basic_blocks.len());
+        for (block, data) in body.basic_blocks.iter_enumerated() {
+            if let TerminatorKind::Call { func, args, destination, .. } = &data.terminator().kind
+                && let [Operand::Move(arg) | Operand::Copy(arg)] = &**args
+                && let Some(arg) = arg.as_local()
+                && destination.as_local().is_some()
+                && let ty::Ref(_, arg_ty, Mutability::Not) = *body.local_decls[arg].ty.kind()
+                && let ty::FnDef(fn_id, _) = *func.ty(body, cx.tcx).kind()
+                && arg_ty.needs_drop(cx.tcx, cx.param_env)
+                && arg_ty.is_freeze(cx.tcx, cx.param_env)
+                && (match_def_path(cx, fn_id, &paths::CLONE_TRAIT_METHOD)
+                    || match_def_path(cx, fn_id, &paths::TO_OWNED_METHOD)
+                    || (match_def_path(cx, fn_id, &paths::TO_STRING_METHOD)
+                        && is_type_lang_item(cx, arg_ty, LangItem::String)))
+            {
+                has_clone = true;
+                clone_calls.insert(block);
+            }
+        }
+        if !has_clone {
+            return;
+        }
 
-        let mir = cx.tcx.optimized_mir(def_id.to_def_id());
-
-        let mut possible_borrower = PossibleBorrowerMap::new(cx, mir);
+        let mut raw_ptr_taken = BitSet::new_empty(body.local_decls.len());
+        for stmt in body.basic_blocks.iter().flat_map(|data| &*data.statements) {
+            if let StatementKind::Assign(a) = &stmt.kind
+                && let Rvalue::AddressOf(_, place) = a.1
+                && !place.is_indirect()
+            {
+                raw_ptr_taken.insert(place.local);
+            }
+        }
 
-        for (bb, bbdata) in mir.basic_blocks.iter_enumerated() {
-            let terminator = bbdata.terminator();
+        let mut ref_target_results = RefTargetAnalysis::new(cx.tcx, cx.param_env, body, &raw_ptr_taken)
+            .into_engine(cx.tcx, body)
+            .iterate_to_fixpoint()
+            .into_results_cursor(body);
+        let clone_sources = clone_calls
+            .iter()
+            .filter_map(|block| {
+                let data = &body.basic_blocks[block];
+                ref_target_results.seek_before_primary_effect(Location {
+                    block,
+                    statement_index: data.statements.len(),
+                });
+                if let TerminatorKind::Call { args, .. } = &data.terminator().kind
+                    && let [Operand::Copy(arg) | Operand::Move(arg), ..] = &**args
+                    && let Some(Some(targets)) = ref_target_results.get().targets.get(&arg.local)
+                {
+                    Some((block, (arg.local, targets.clone())))
+                } else {
+                    None
+                }
+            })
+            .collect::<FxHashMap<_, _>>();
```

Comment:
Typically yes. If there ends up being perf problem this would be worth not linting the few extra cases this could catch. Limiting the analysis to just types which are cloned could also get a decent speed up